microsoft/PowerToys
# Edit Shortcuts UI (dev/keyboardManager)

## Summary of the Pull Request
Added the Edit Shortcuts UI and reformatted the code to include a common library which has a state class that can be used to share information between the UI and the back-end.

## Details of the Pull Request
- Added a common library with Helper functions and KeyboardManagerState class. This class will be used to share information between the hook and the UI. The object of this class is created in dllmain.cpp.
- Refactored the code base to constrain all code which involves UI and back-end information to wrapper functions in the state class.
- Added a shortcutControl class which creates a custom control for each element in the shortcut table.

## Follow up To-Dos
* Add thread safety
* Add tests
* Add support to save the shortcuts to a file
* Add UI for app specific shortcuts and configurations
* Display key names instead of codes


## PR Checklist
* [X] Applies to #6 


## Validation Steps Performed
- Run PowerToys.exe. The Keyboard Manager window should pop up if PowerKeys is enabled.
- Click Edit Shortcuts. Clicking the + Sign will allow you to add a new shortcut. The shortcut can be removed by clicking the delete button.
- Press Type shortcut to detect your shortcut. As of now the key name will not be displayed, only the code.
- In the current version, only 2 key shortcuts will be applied. A check has not been added for the order of the keys in the UI, so while testing it make sure the modifier key is clicked first. Example: Ctrl+A and not A+Ctrl.
- Clicking Apply will save the settings and your shortcuts will be remapped. If any of remaps were invalid, the valid ones will still apply, however a warning will be displayed on the top right corner.
- Clicking Cancel will simply exit the window. If Apply was not clicked prior to clicking cancel then those shortcuts will not be saved.
- If you reopen the Edit Shortcuts Window your shortcuts will still be visible. However if you terminate PowerToys.exe and relaunch it your saved shortcuts will be lost.

![editshortcuts](https://user-images.githubusercontent.com/32061677/77211389-16aeac80-6ac1-11ea-9ae0-4d60669c1445.gif)
